07/scala/homework07/src/main/scala/example/TypeClassesHomework.scala

```diff
@@ -0,0 +1,150 @@
+    implicit class ShowOps[T](x: T) {
+      def show(implicit showable: Show[T]): String = showable.show(x)
+    }
+  }
+
+  final case class User(id: String, name: String)
+
+  // Exercise: create Show instance for User (interface implementation)
+  implicit val showUser: Show[User] = user => s"id: ${user.id}; name: ${user.name}"
+
+  import ShowSyntax._
+  println(User("1", "Oleg").show)
+}
+
+object Task3 extends App {
+  type Error = String
+  trait Parse[T] { // invent any format you want or it can be csv string
+    def parse(entity: String): Either[Error, T]
+  }
+
+  // Exercise: create syntax for Parse so i can do "lalala".parse[User] (and get an error because it is obviously not a User)
+  object ParseSyntax {
+    implicit class ParseOps[T](x: String) {
+      def parse[T](implicit parseable: Parse[T]): Either[Error, T] = parseable.parse(x)
+    }
+  }
+
+  final case class User(id: String, name: String)
+
+  // Exercise: create Parse instance for User
+  implicit val parseUser: Parse[User] = s => s.split("""\s+""") match {
+    case Array(id, name) => Right(User(id, name))
+    case _ => Left ("Invalid user string")
+  }
+
+  import ParseSyntax._
+  println("lalala".parse[User])
+  println("someId someName".parse[User])
+}
+
+object Task4 extends App {
+  // Exercise: design a typesafe equals so i can do a === b, but it won't compile if a and b are of different types
+  // define the typeclass (think of a method signature)
+  // remember `a method b` is `a.method(b)`
+  trait Equals[T] {
+    def equals(e1: T, e2: T): Boolean
+  }
+
+  object EqualsSyntax {
+    implicit class EqualsOps[T](e1: T) {
+      def ===(e2: T)(implicit equatable: Equals[T]): Boolean = equatable.equals(e1, e2)
+    }
+  }
+
+  final case class User(id: String, name: String)
+  final case class SuperUser(id: String, name: String)
+
+  implicit val userEquals: Equals[User] = (e1, e2) => e1.id == e2.id && e1.name == e2.name
+
+  import EqualsSyntax._
+  println(User("1", "Oleg") === User("1", "Oleg"))
+  // println(User("1", "Oleg") === SuperUser("1", "Oleg")) // Will not compile (Type mismatch: Required: User; Found: SuperUser)
+}
+
+object AdvancedHomework extends App {
+  // Exercise: create a typeclass for flatMap method
+  // Considering there is a flatMap in Scala, I did implementation using F# naming - bind
+  trait Bind[E[_]] {
+    def bind[T](eT: E[T], f: T => E[T]): E[T]
+  }
+
+  object BindSyntax {
+    implicit class BindOps[T, E[_]](eT: E[T]) {
+      def bind(f: T => E[T])(implicit bindable: Bind[E]): E[T] = bindable.bind(eT, f)
+    }
+  }
+
+  implicit val optionBind: Bind[Option] = new Bind[Option] {
+    def bind[T](eT: Option[T], f: T => Option[T]): Option[T] = eT match {
+      case Some(t) => f(t)
+      case _ => None
+    }
+  }
+
+  implicit val listBind: Bind[List] = new Bind[List] {
+    def bind[T](eT: List[T], f: T => List[T]): List[T] = eT match {
+      case head :: tail => f(head).concat(bind(tail, f))
```

Comment:
Im not sure i know how concat is defined but we probably can write it as `f(head) ::: bind(tail, f)`

Reply:
Works exactly the same way. Thanks!